### DISCUSSION TITLE
Subscribe to Langfuse Events via Webhooks

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#1033] [Ideas]
Langfuse produces many different events. For various reasons it can be useful to subscribe to them via a webhook.

Potential events
- score:create (user feedback, eval, ...)
- trace:create
- prompt:create

Very early idea, raised by a couple of teams using Langfuse. Feel free to add your thoughts or what you'd like to see/contribute.

-----
**Edit**

We have a first webhook version out there: https://langfuse.com/docs/prompt-management/features/webhooks-slack-integrations

If you have any questions or feedback, please let us know! Sending out webhooks on prompt versions is only the start. We will gradually increase the scope of the feature.

----

**Update - June 18th, 2026**

Our new [Alerts](https://cloud.langfuse.com/project/~/alerts) feature just shipped! As requested by many of you, alerts are triggered based on aggregated metrics over a sliding window. [Learn more in the docs](https://langfuse.com/docs/metrics/features/alerts).

If you have any questions or feedback, please let us know.

**Comment by khareyash05** [comment]
Hey This feature will be very useful! Will this be a Slack Webhook Url or a Discord Webhook Url?

**Reply by marcklingen** [reply]
Currently thinking about generic webhook. It is easy to then setup a Slack workflow to consume this information and print to Slack or use Zapier/Make/n8n to build a quick custom consumer in order to pipe the events to other channels.

**Reply by khareyash05** [reply]
Hey another event I can think of is whenever a new member is added/removed to/from the project 
Also with webhook url we can also send an email that this person was added to the workspace by this person.

**Comment by clemra** [comment]
+ 1 in #1512

**Reply by hypernovas** [reply]
Thanks! Curious what is the status for the webhook capability? Something we were looking at is based on the cost burst, LLM API outage or latency spikes, we can publish events. This can be very useful for LLM app ops for many orgs.

**Reply by marcklingen** [reply]
We are waiting for v3 changes before adding webhook capabilities. Thereby these can happen behind async queues instead of having to be run within sync apis. More context on this here: https://github.com/orgs/langfuse/discussions/1902

**Comment by chrisshaw** [comment]
Sorry to bother twice, walking over from #1512 . Wanted to extend the [very good use cases we also have](https://github.com/orgs/langfuse/discussions/1033#discussioncomment-8933260) that @hypernovas brought up with what I shared in [my reply](https://github.com/orgs/langfuse/discussions/1512#discussioncomment-9757098) in the prior discussion: alerts based on sliding window aggregated scores. Responding to events from _individual_ scores, traces, etc seems fun but not urgent. In our case, we want to know that the moderation LLM app is still moderating correctly and want to be alerted based on an aggregation of scores crossing a threshold.

**Reply by marcklingen** [reply]
thanks for the feedback and sorry for the slow response. I fully agree that alerts are very distinct from rule-based triggers on every newly created object in Langfuse. I create a new thread to track this independently: https://github.com/orgs/langfuse/discussions/3997

**Reply by ahmed-moubtahij** [reply]
>  alerts based on sliding window aggregated scores

+1 for this need

**Comment by ajram23** [comment]
+1 please.

**Comment by arackley-tribal** [comment]
+1 Would love to be able to use this to be notified when Langfuse is noticing errors in traces etc.

**Comment by barseghyanartur** [comment]
Copied from https://github.com/orgs/langfuse/discussions/4342

----

**Feature request: Lifecycle hooks in LangFuse**

Introduce lifecycle hooks in LangFuse to trigger custom actions during key events (such as prompt promotion to production).

These hooks should provide flexibility, allowing integration with external systems via webhooks.

Initial implementation could be as simple as managing that through the UI:

- Event to hook into.
- Webhook URL.
- Configuration (security tokens, retries, timeouts).
- Additional payload parameters.

On a long term, support for a reproducible configuration format for defining lifecycle hooks would be useful. Think of version-controlled deployments or "manual" environment duplications.

----

**Reply by maxdeichmann** [reply]
@barseghyanartur thanks a lot for sharing this! Can you maybe share a bit which events exactly would be helpful for you and what you would do with them?

**Reply by barseghyanartur** [reply]
Thanks for getting back.

I'd love as many hookable events as possible, but these three are my highest priority:

#### Prompt label created

**Event**: Prompt label created.  
Fires whenever a prompt receives a (new) label. For example, I have a suite of integration tests that should run whenever prompts change. I can define custom labels like `dev` or `qa` (in addition to `production`) and automatically verify that everything still works after an update.

#### Prompt label removed

**Event**: Prompt label removed.  
Fires whenever an existing label is removed. If someone accidentally removed the `production` label, my integration tests will catch it immediately - preventive maintenance rules!

> *I expect this to cover the case where a prompt itself is removed.*

#### Prompt updated

**Event**: Prompt updated.  
Fires whenever the prompt text or its configuration changes. For instance, if someone tweaks the user- or system-prompt text or renames a variable, integration tests are executed.

**Reply by maxdeichmann** [reply]
Awesome, thanks so much. We have a PR for exactly these events and will launch this asap. I will keep you posted here.

**Comment by chai3** [comment]
datasets:create
I want to run agent experiments(not prompt experiments)

**Reply by chai3** [reply]
Thanks
https://langfuse.com/changelog/2025-07-24-remote-experiment-triggers

**Comment by camwardy** [comment]
prompt:create would be really useful for our team - we'd like to store the names in our DB

**Reply by maxdeichmann** [reply]
@camwardy thanks a lot for the feedback! What exactly would be your workflow here if you had such a webhook?

As side note: you are able to configure refetching new prompts in the SDK ([docs](https://langfuse.com/docs/prompts/get-started#caching-in-client-sdks)).

**Comment by george-bsq** [comment]
+1 would be useful when the production label moves to a different version of the same prompt.

**Reply by marcklingen** [reply]
thanks for sharing

**Reply by maxdeichmann** [reply]
@george-bsq thanks a lot for the feedback! What exactly would be your workflow here if you had such a webhook?

As side note: you are able to configure refetching new prompts in the SDK ([docs](https://langfuse.com/docs/prompts/get-started#caching-in-client-sdks)).

**Reply by george-bsq** [reply]
@maxdeichmann Hi Max, thank you for showing interest in this.

While I'd love to use Langfuse as my prompt manager, I'd prefer not to add another third-party dependency to my system, let alone one that my system initialization depends on. Given a hook like the one I described, I could use it to deploy the prompt to my own infrastructure whenever a new version is promoted to 'production'.

**Reply by maxdeichmann** [reply]
@george-bsq Awesome, thanks so much. We have a PR for exactly these events and will launch this asap. I will keep you posted here.

**Comment by mohamedgomran** [comment]
`prompt:update` will also be useful, the langfuse SDK is caching the prompt in the the memory and we need a way to get notified when prompt changes so we can invalidate the cache

**Reply by lasryaric** [reply]
Absolutely, that's why I ended up here. 
Our use case:
We are building a pretty elaborated AI agent (for dating :)) and we need a super low latency, with always up to date prompts in order for our product team to iterate as quickly as possible.

**Reply by marcklingen** [reply]
> the langfuse SDK is caching the prompt in the the memory

You can configure the cache TTL. The SDK will deliver the stale prompt to optimize latency and revalidate in the background. By default, this occurs every minute. Does this not work for you?

https://langfuse.com/docs/prompts/get-started#caching-in-client-sdks

**Reply by ajram23** [reply]
Can confirm that this works for me and it really valuable. @lasryaric highly recommend it!

**Comment by AlmogBaku** [comment]
any updates on that?

**Comment by lasryaric** [comment]
A ttl is not enough because I want to have caching for the prompts (they
take up to 300ms to fetch) but I also want to be able to deploy new prompts
instantly by invalidating my cache. Hence the need for the webhook.


On Tue, Mar 18, 2025 at 9:54 AM Almog Baku ***@***.***> wrote:

> any updates on that?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/langfuse/discussions/1033#discussioncomment-12535394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACXUNPEEIQPEPQL5NORPTT2U7NN3AVCNFSM6AAAAABCVNXDZOVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENJTGUZTSNA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Aric Lasry
+33 7 49 50 55 74

**Comment by agbragin** [comment]
+1 `trace:create` and `score:create` would be extremely useful for automatic triggering of custom evaluation pipelines.

**Reply by maxdeichmann** [reply]
Hi @agbragin thanks a lot for sharing this! What is the use-case for this? Does the open-sourcing of LLM as a judge in Langfuse help you here to run evals async?

**Reply by agbragin** [reply]
Hi! Thanks for the response! I would say that open-sourcing of LLM-as-a-judge would _not_ solve the issue. 

The use case is to evaluate traces (including the complex ones, e.g. agentic) on arrival using all kinds of evaluation (functional, heuristics-based, gold comparison).

`score:create` can be used to push notification when scores are out of the desired range.

**Reply by Haoping-Xiao** [reply]
same here. @maxdeichmann 
1. Our local host LLM can't support structure output and langfuse evalutor couldn't work.  `trace:create` can help this.
2. Evaluation might need processing in advance. E.g: trajectory evaluation and using metadata to help understand the task.
3. `score:create` help push notification.

**Reply by maxdeichmann** [reply]
@agbragin + @Haoping-Xiao: we thought to also have `score-create` and `trace-create` events. Yet, if your application has very high load, this would also mean that we push out many webhooks to your systems. Would you be willing to deal with this load or would you like to have some sort of sampling / rate limiting?

**Reply by Haoping-Xiao** [reply]
I think a general rate limiting is convenient. In addition, filter settings for traces/scores[threshold settings] are also helpful to reduce the load.

**Reply by agbragin** [reply]
@marcklingen Currently rate limiting is not critical for us – the idea is to configure triggers for selected projects only (offline evaluation).

**Comment by maxdeichmann** [comment]
Hi everyone,

we are currently actively looking into this feature! I would love to learn a bit more which use-cases might be helpful here. So far I think most requested are alerting based on aggregates (e.g. score value hits threshold over time) or webhooks once prompts change (e.g. new version / label). Do you have any other use cases?

If yes, tell me more about what data you need to run your workflows and I will happily consider that.

**Reply by luis-seibert** [reply]
**I'm very much looking forward to this feature!**

I can see myself using it for reporting exceptions.
However, what I was missing for using Langfuse for that was:

- the ability to turn off the automatic error level setting (as this interferes with manual settings).
- the ability to set error levels at trace level and not just observation level.
- more or customisable level tags to allow more granular control over this.

Then I think this could be really useful as a webhook to report errors to Slack, for example.

**Reply by ajram23** [reply]
Here are my basic cases 1. If a run fails, I'd like to log a GitHub ticket 2. If the score is under a certain value I'd like to log a ticket.

**Reply by robertkirk13** [reply]
We use webhooks to update a cache of the prompts in a KV connected to our serverless backend (sdk caching doesn't work because it's a new instance every time). This is something we need to switch to langfuse.

**Reply by jcbmllgn** [reply]
Hello, we're excited for this feature as well. In addition to the needs that many others have surfaced, we'd also like webhooks for when a metric defined on a dashboard falls below some threshold. We're specifically looking to do alerting if an Eval falls below a threshold.

**Reply by Haoping-Xiao** [reply]
- Our local host LLM can't support structure output and langfuse evalutor couldn't work. Can we trigger a `eval:create` signal when a custom evaluation pipeline setup is statisfied? This would be phenomenal. Otherwise, a `trace:create` signal that with some filters is great as well. `eval:create` can help manage the eval config while `trace:create` is not intuitive.
- Evaluation might need processing in advance. E.g: trajectory evaluation and using metadata to help understand the task.
- score:create help push notification.

**Reply by Haoping-Xiao** [reply]
> * Our local host LLM can't support structure output and langfuse evalutor couldn't work. Can we trigger a `eval:create` signal when a custom evaluation pipeline setup is statisfied? This would be phenomenal. Otherwise, a `trace:create` signal that with some filters is great as well. `eval:create` can help manage the eval config while `trace:create` is not intuitive.
> * Evaluation might need processing in advance. E.g: trajectory evaluation and using metadata to help understand the task.
> * score:create help push notification.

https://docs.smith.langchain.com/observability/how_to_guides/online_evaluations#get-started-with-online-evaluators

I think langsmith custom code evaluator is a great feature. while they intergrated it on the platform, it would make sense to have a trigger to call my custom evaluation pipeline.

**Reply by maxdeichmann** [reply]
Thanks, everyone for coming back here! We will use this to educate a V0 and all follow up additions we make to this feature. We will start out with webhooks for prompt changes but add more over time.

**Comment by kyu-aurelian** [comment]
I would love to have this feature supported.
basically, I would like to be notified if there any prompt label changes etc, as they are essneitally prod deployment.

**Comment by romusters** [comment]
I'd like to have an event, that when the prompt label changes, a configurable endpoint is called where the label is passed together with the prompt version. This way, I can call an endpoint to run custom evaluation metrics and/or afterwards trigger a Github action. 

Currently the process is manual and a lot of responsibility is put on a single colleague who presses the "production" button in Langfuse. The button press impacts production directly without having a automated quality gate/process.

**Comment by maxdeichmann** [comment]
Hi all, we have a first webhook version out there: https://langfuse.com/docs/prompts/get-started#webhooks

If you have any questions or feedback, please let us know! Sending out webhooks on prompt versions is only the start. We will gradually increase the scope of the feature.

**Reply by kyu-aurelian** [reply]
awesome!
few feedback/questions:
- I see only a single option for `Name` in the filter. I guess there is a plan to add more columns like `Tags`, `Labels` etc?
- Could you also add a feature to be able to customize webhook payload? I tried to integrate it with Slack Webhook API, but was not able to as there was no way to customize webhook payload to make it conform the schema Slack expects. 
  - I understand it can be handled if we had our own webhook handler but would ideally be better if there is a direct way to achieve the same.

**Reply by maxdeichmann** [reply]
Hi @kyu-aurelian, thanks a lot for the feedback.

- For the filters: yes, we could add more. What exactly would you need them for?
- Adjusting the payload: yes this would be great but I fear that this would make things complicated in the UI. We will for sure add a native slack integration.

**Reply by kyu-aurelian** [reply]
for filters, i would like to be notified on label changes (e.g. if a newer version of prompt is promoted to prod by having production label).
we basically want to be notified when there are changes that can affect production environment. 

and yes. it doesn't have to be supporting custom payload. if native slack integration can be supported that would be great!
fwiw, I have seen a few services use `https://www.svix.com/` for their webhook integrations, which I thought was quite convenient.

**Reply by vardior9** [reply]
@maxdeichmann I see that the current Prompt webhook code doesn't lay an infrastructue for additional event support - is this something worth checking out myself or are you guys in the middle of implementation and it's worth waiting for the next phase before extending it with my use case (evaluation score alerts)?

**Reply by maxdeichmann** [reply]
Hi @vardior9 thanks a lot for raising this question. We have this on our road map. However, as you noticed, we do not have the infra yet to support score based alerts. As this is extremely core to our infra, we will build this within our team as we know best which usage patterns of clickhouse will work.

**Comment by paolotamag** [comment]
can this be used to alert us on slack when langfuse traces are failing (LEVEL : ERROR)?

**Comment by vaibhav3002** [comment]
@marcklingen  - Hello, could you please tell the availability of the event subscription support via webhooks for scores as well?

**Comment by maxdeichmann** [comment]
Hi all,

thanks so much for the interest in Langfuse and sharing your needs here! 

We launched the slack integration for prompt events recently (https://langfuse.com/changelog/2025-07-30-slack-prompt-notifications).

I know that many of you want to have tracing based alerts (e.g. score values, cost, latencies..) for these integrations. We have this very high up on our roadmap. We need to complete some more infrastructure work in the background to enable such alerts at scale unfortunately.

**Comment by amansr09** [comment]
We are using LLM-as-a-Judge for automated evaluations and need to be alerted in real-time if a trace evaluation is incorrect, as manually checking the UI is not scalable for our production pipeline. can we get a new webhook event that triggers when an evaluation score falls below a set threshold would be a great solution, allowing us to send an immediate Slack notification and investigate.

**Reply by ezou-shiba** [reply]
+1

**Reply by vrinda-squareup** [reply]
Hi @marcklingen - this is something we would love seeing as well. Webhooks when eval job is completed / failed. Let us know if this is already being built

**Reply by antoinehirtz-devoteam** [reply]
+1

**Comment by jkryanchou** [comment]
I need a webhook or any other subscribed way to trigger next action. like trigger a webhook notification or call the n8n workflow.

**Comment by BrendanRitter** [comment]
Our team is using langfuse for catching eval failures and doing x% general review. It would be great if you could tie a annotation queue to alerting so we knew when items have been added to those queues.

**Comment by manuel-reil** [comment]
We want to track token consumption, cost via openmeter.io for billing purposes. So webhook for the respected events would be good.

**Comment by ahmed-moubtahij** [comment]
+1 to alerts on trace latency and +1 to alerts on sliding window aggregated eval scores

**Reply by howdymatt** [reply]
Very important to us too

**Reply by GNamDor** [reply]
important to us too

**Reply by marco-kalipso** [reply]
same here!

**Reply by bostrt** [reply]
Same

**Reply by tswann** [reply]
Same

**Comment by musaabhasan** [comment]
For webhooks tied to traces, scores, prompts, and annotation queues, I would design for replay and idempotency from the start. These events often drive downstream incident workflows, billing, or evaluation review, so duplicate delivery should be safe and missed delivery should be recoverable.

Fields I would want in each webhook envelope:

- event_id, event_type, event_version
- project_id and environment
- resource_type and resource_id
- created_at and emitted_at
- trace_id / observation_id / prompt_version where relevant
- actor type, if available, without exposing unnecessary user data
- signature header and timestamp for verification
- delivery attempt number

For eval/annotation workflows, it would also help to include threshold context: score name, score value, threshold, baseline window, and reason the event fired. That lets teams route “latency regression”, “faithfulness regression”, and “human review queue item created” differently instead of treating every webhook as a generic notification.

**Comment by marksalpeter** [comment]
Hi everyone! Our new [Monitors](https://cloud.langfuse.com/project/~/monitors) feature just shipped in cloud beta. You can set threshold alerts on metrics like cost, scores, and latency and route them to Slack, webhooks, or GitHub Actions ([docs](https://langfuse.com/docs/metrics/features/monitors)).

If you have any questions or feedback, please let us know.

cc: @hypernovas, @chrisshaw, @arackley-tribal, @jcbmllgn, @paolotamag, @vardior9, @amansr09, @ahmed-moubtahij, @manuel-reil, @howdymatt, @GNamDor, @marco-kalipso, @bostrt, @tswann

**Reply by ahmed-moubtahij** [reply]
Thank you! We will definitely be trying this

**Reply by marued** [reply]
YES!

**Reply by hhhhhhhhhhhhhhhhho** [reply]
Thanks for this feature!

I tried the monitoring feature just a moment ago, and it was very useful for my usecase.

I'm looking forward to seeing in production release.

Do you have an estimated timeline for when it will be available in the self-hosted production release?

**Reply by marksalpeter** [reply]
Hi @hhhhhhhhhhhhhhhhho! This will be released for self hosters with v4 enabled soon. For now, we're still stress testing it on cloud.
